libbpfgo/helpers/kernel_features.go

```diff
@@ -0,0 +1,110 @@
+package helpers
+
+import (
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"golang.org/x/sys/unix"
+)
+
+type KernelConfig map[string]string
+
+// NewKernelConfig simply creates the underlying map used by
+// the KernelConfig, you must use InitKernelConfig to read
+// in the values of the config
+func NewKernelConfig() KernelConfig {
+	return make(map[string]string)
+}
+
+// InitKernelConfig attempts to read the kernel config
+//  into the KernelConfig from either
+// - /proc/config-$(uname -r)
+// - /boot/config.gz
+func (k KernelConfig) InitKernelConfig() error {
+
+	err := k.getBootConfig()
+	if err == nil {
+		return nil
+	}
+
+	err2 := k.getProcGZConfig()
+	if err != nil {
+		return fmt.Errorf("%v %v", err, err2)
+	}
+
+	return nil
+}
+
+// GetKernelConfigValue retrieves a value from the kernel config
+// If the config value does not exist an error will be returned
+func (k KernelConfig) GetKernelConfigValue(key string) (string, error) {
+	v, exists := k[key]
+	if !exists {
+		return "", errors.New("kernel config value does not exist, it is either not known by your kernel version or you have not called InitKernelConfig")
```

Comment:
I think that instead of forcing the user to call InitKernelConfig, we should use "call once". By using it, InitKernelConfig (which can then be not exported) will be called only once at the first time this function is called.
(I remember that go supports that in some way, I can find you a reference if you want).

Reply:
See my response to Itay above,  combining New/Init I think works best.

Reply:
Golang once (from the sync package) is not the same as using init. It is only called the first time you call this getter, and not when the package is initialized

Reply:
Oh interesting! Checking it out

Reply:
> > this is great functionality to have in libbpfgo. May I suggest some UX improvements:
> > Instead of:
> > ```go
> > import github.com/.../libbpfgo/helpers
> > ...
> > helpers.NewKernelConfig()
> > helpers.InitKernelConfig()
> > conf := helpers.KernelConfig["something"]
> > ```
> > 
> > 
> > we can have:
> > ```go
> > import github.com/.../libbpfgo/kconfig
> > ...
> > conf, err := kconfig.Get("something)
> > ```
> > 
> > 
> > I don't see why a user would care to instantiate the map, then populate it manually every time they want to get a config. Instead we can just do all of that in the package's `init()`. WDYT?
> 
> That's how I originally designed it (with an init) but decided against it. I'm not sure _why_ this would happen, but I was worried about a situation where the config file is created after the init function is called. I suppose this could happen if the kernel is compiled with `CONFIG_IKCONFIG` but not `CONFIG_IKCONFIG_PROC` and the file is extracted to /proc after the fact (again, not sure why that'd happen). I'd say in general I found the design to be more dangerous for suppressing errors that can't be repaired.
> 
> I will say, there's no reason to have both a `NewKernelConfig` and an `InitKernelConfig`, I am going to put the logic for Init into New.

I would personally vote for a no init solution. While on the surface it seems appealing by the user having to do less setup work, it is a little [_magical_](https://peter.bourgon.org/blog/2017/06/09/theory-of-modern-go.html). 

How about the following:
```
var kc KernelConfig
conf, err := helpers.NewKernelConfig(kc)
...
val, err := conf. GetKernelConfigValue("some key")
```

This enables the user to inject their own KernelConfig type which gets modified by the constructor.

Reply:
I feel this is clever but wouldn't be the expected behavior of someone using this API so it's best avoided. Having them just make one call to `InitKernelConfig` is acceptable now that I merged New and Init.